### PR TITLE
[Examples] RGCN Heterogeneous on ogbn-mag

### PR DESCRIPTION
**Pull request by ksadowski13**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
This is implementation of RGCN Heterogenous based on RGCN Homogeneous version which learns well on ogbn-mag (~47% accuracy which matches OGB leaderboard submission).

Additional features:
- Model implementation works with mini-batch and full-graph training 
- 3 inference modes: neighbor-sampler, full-neighbor-sampler and full-graph
- Early stopping callback that saves model parameters from best epoch and use them for test validation
- Input dropout
- Possibility to use batch norm
- Possibility to set different norm for adjacency matrix for RGCN layers

Things that differs from RGCN Hetero from current examples:
- Embedding layer similar to homogeneous version with possibility to project node features that are present in graph into different size

Things that differ from RGCN Homogeneous implementation:
- Distributed training isn't implemented

Benchmarks on AWS g4dn.8xlarge:
train_device | inference_device | inference_mode | num_layers | hidden_feats | batch_size | eval_batch_size | fanouts | avg_train_time | avg_valid_time | avg_full_time | test_accuracy
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
gpu | gpu | neighbor_sampler | 2 | 64 | 1024 | 1024 | 25,20 | 39.2500 | 3.3782 | 42.6849 | 0.4697
gpu | gpu | neighbor_sampler | 2 | 64 | 1024 | 8192 | 25,20 | 38.8948 | 2.6740 | 41.6316 | 0.4726
gpu | gpu | neighbor_sampler | 2 | 64 | 1024 | 16384 | 25,20 | 38.9087 | 2.3495 | 41.3256 | 0.4786
gpu | gpu | full_neighbor_sampler | 2 | 64 | 1024 | 1024 | 25,20 | 39.0707 | 38.5087 | 77.6348 | 0.4696
gpu | gpu | full_neighbor_sampler | 2 | 64 | 1024 | 8192 | 25,20 | 39.1360 | 9.7092 | 48.9007 | 0.4694
gpu | gpu | full_neighbor_sampler | 2 | 64 | 1024 | 16384 | 25,20 | 39.0038 | 8.7021 | 47.7617 | 0.4569
gpu | cpu | full_graph | 2 | 64 | 1024 | nan | 25,20 | 39.4855 | 4.1055 | 44.2090 | 0.4747
gpu | gpu | neighbor_sampler | 3 | 256 | 1024 | 1024 | 5,10,15 | 134.3128 | 10.5660 | 145.0134 | 0.4724
gpu | gpu | neighbor_sampler | 3 | 256 | 1024 | 8192 | 5,10,15 | 134.6823 | 6.8027 | 141.5735 | 0.4707
gpu | gpu | neighbor_sampler | 3 | 256 | 1024 | 16384 | 5,10,15 | 134.8789 | 4.9968 | 139.9730 | 0.4712
gpu | gpu | full_neighbor_sampler | 3 | 256 | 1024 | 1024 | 5,10,15 | 134.2004 | 57.6302 | 191.9009 | 0.4863
gpu | gpu | full_neighbor_sampler | 3 | 256 | 1024 | 8192 | 5,10,15 | 133.9809 | 14.9223 | 148.9736 | 0.4843
gpu | gpu | full_neighbor_sampler | 3 | 256 | 1024 | 16384 | 5,10,15 | 134.7735 | 13.4857 | 148.3300 | 0.4901
gpu | cpu | full_graph | 3 | 256 | 1024 | nan | 5,10,15 | 133.7911 | 10.7073 | 145.2222 | 0.4640

Training on GPU and inference on GPU using `MultiLayerNeighborSampler`:
- `python main.py --inference-mode neighbor_sampler`

Training on GPU and inference on GPU using `MultiLayerFullNeighborSampler`:
- `python main.py --inference-mode full_neighbor_sampler`

Training on GPU and inference on CPU with full-graph:
- `python main.py --no-gpu-inference --inference-mode full_graph`

### CODE REVIEW

File: examples/pytorch/rgcn-hetero-ogbn-mag/model.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        self,
+        in_nodes: Dict[str, torch.Tensor] = None,
+        device: torch.device = None,
+    ) -> Dict[str, torch.Tensor]:
+        if in_nodes is not None:
+            x = {}
+
+            for ntype, nid in in_nodes.items():
+                if self._node_feats[ntype] is None:
+                    x[ntype] = self.node_embeddings[ntype](nid)
```
[the remark this answers is not in the document]
Comment (ksadowski13) [reply examples/pytorch/rgcn-hetero-ogbn-mag/model.py]: Done in this [commit](https://github.com/dmlc/dgl/pull/3371/commits/3fc78f012ecb6206cbc801683fed76bb2fb51d03). I've moved `if else` statements before `for loop`.

**Comment by ksadowski13** [comment]
The code comes from my experiments with full-graph vs mini-batch convergence of some models and I mentioned this implementation on our sync with Da, so he asked to make PR to check if this is interesting for DGL team. If you find this implementation useful I could work on this further to unify it with other DGL examples or extend this by adding more datasets and documentation.

File: examples/pytorch/rgcn-hetero-ogbn-mag/model.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            if hg.is_block:
+                inputs_dst = {ntype: h[:hg.num_dst_nodes(
+                    ntype)] for ntype, h in inputs.items()}
+            else:
+                inputs_dst = inputs
+        else:
+            inputs_dst = None
+
+        x = self._conv(hg, inputs, mod_kwargs=weight_dict)
+        x = {ntype: self._apply_layers(ntype, h, inputs_dst)
```
[the remark this answers is not in the document]
Comment (ksadowski13) [reply examples/pytorch/rgcn-hetero-ogbn-mag/model.py:164]: Do you mean the situation when destination nodes don't have any neighbors? If yes, should there be `try except` block for checking this?

File: examples/pytorch/rgcn-hetero-ogbn-mag/main.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+            hg,
+            {predict_category: valid_idx},
+            valid_sampler,
+            batch_size=args.eval_batch_size,
+            shuffle=False,
+            drop_last=False,
+            num_workers=args.eval_num_workers,
+        )
+
+        if args.test_validation:
```
[the remark this answers is not in the document]
Comment (ksadowski13) [reply examples/pytorch/rgcn-hetero-ogbn-mag/main.py:198]: This argument was used for my hyperparameter optimization experiments where optimized metric was loss on valid dataset, so I didn't need test evaluation in this phase. By default it's set to true, but I could just remove it if you find it more appropriate way.

**Review by mufeili — APPROVED** [review]
The code is of high quality. Thank you for the great work.

**Comment by mufeili** [comment]
@ksadowski13 We may need to perform some re-organization and modification as needed in the future. If you are fine with it, we can merge this PR first.

**Comment by ksadowski13** [comment]
@mufeili Sure, I will be happy to work on this further, so let me know if I could help in the future with modifications :)
